lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
       labels
     end
 
-    # Determine the group level monitored resource and common labels shared by a
-    # collection of entries.
-    def determine_group_level_monitored_resource_and_labels(tag)
-      # Determine group level monitored resource type. For certain types,
-      # extract useful info from the tag and store those in
-      # matched_regex_group.
-      group_resource_type, matched_regex_group =
-        determine_group_level_monitored_resource_type(tag)
-
-      # Determine group level monitored resource labels and common labels.
-      group_resource_labels, group_common_labels =
-        determine_group_level_labels(group_resource_type, matched_regex_group)
-
-      group_resource = Google::Apis::LoggingV2beta1::MonitoredResource.new(
-        type: group_resource_type,
-        labels: group_resource_labels.to_h
-      )
-
-      # Freeze the per-request state. Any further changes must be made on a
-      # per-entry basis.
-      group_resource.freeze
-      group_resource.labels.freeze
-      group_common_labels.freeze
-
-      [group_resource, group_common_labels]
+    # Group the log entries by tag and locally_unique_id pairs.
+    def group_log_entries_by_tag_and_locally_unique_id(chunk)
+      groups = {}
+      chunk.msgpack_each do |tag, time, record|
+        sanitized_tag = sanitize_tag(tag)
+        if sanitized_tag.nil?
+          @log.warn "Dropping log entries with invalid tag: '#{tag}'. " \
+                    'A tag should be a string with utf8 characters.'
+          next
+        end
+        locally_unique_id = record.delete(DEFAULT_UNIQUE_ID_KEY) if
+          record.is_a?(Hash)
+        hash_key = [sanitized_tag, locally_unique_id].freeze
+        groups[hash_key] ||= []
+        groups[hash_key].push([time, record])
+      end
+      groups
     end
 
-    # Determine group level monitored resource type shared by a collection of
-    # entries.
-    # Return the resource type and tag regexp matched groups. The matched groups
-    # only apply to some resource types. Return nil if not applicable or if
-    # there is no match.
-    def determine_group_level_monitored_resource_type(tag)
+    # Determine the group level monitored resource and common labels shared by a
+    # collection of entries.
+    def determine_group_level_monitored_resource_and_labels(tag,
+                                                            locally_unique_id)
+      resource = @resource.dup
+      resource.labels = @resource.labels.dup
+      common_labels = @common_labels.dup
+      matched_regexp_group = nil
```

Comment:
Yes. if it's not `GKE`, the `@tag_regexp_list` will be empty. I added this line because some tests were failing.

Reply:
How would you feel about calling it out explicitly with a "hideously awesome" recipe I [found somewhere](https://stackoverflow.com/a/13299030/6752892):
```ruby
      if @tag_regexp_list.each do |derived_type, tag_regexp|
        matched_regexp_group = tag_regexp.match(tag)
        if matched_regexp_group
          resource.type = derived_type
          break
        end
      end.empty?
        matched_regexp_group = nil
      end
```
? Or with something less aw{esome,ful} and more conventional, e.g.:
```ruby
      unless @tag_regexp_list.empty?
        @tag_regexp_list.each do |derived_type, tag_regexp|
          matched_regexp_group = tag_regexp.match(tag)
          if matched_regexp_group
            resource.type = derived_type
            break
          end
        end
      else
        matched_regexp_group = nil
      end
```
?

Reply:
The first one kind of scared me...

We would need to initiate `matched_regexp_group` in the beginning because of how ruby decides its variable scope. As shown below `if` works, but `for each loop` does not.

```
$ irb
2.1.10 :002 > if true
2.1.10 :003?>     foo = 'a'
2.1.10 :004?>   else
2.1.10 :005 >       foo = 'b'
2.1.10 :006?>   end
 => "a"
2.1.10 :007 > foo
 => "a"
2.1.10 :008 > {'a' => 'a'}.each do |k, v|
2.1.10 :009 >       val = 'val'
2.1.10 :010?>   end
 => {"a"=>"a"}
2.1.10 :011 > val
NameError: undefined local variable or method `val' for main:Object
        from (irb):11
        from /usr/local/google/home/lingshi/.rvm/rubies/ruby-2.1.10/bin/irb:11:in `<main>'
```

I changed the code to be
```
      matched_regexp_group = nil
      # @tag_regexp_list can be an empty list.
      @tag_regexp_list.each do |derived_type, tag_regexp|
        matched_regexp_group = tag_regexp.match(tag)
        if matched_regexp_group
          resource.type = derived_type
          break
        end
      end
```
if that helps.

Reply:
Ah, I didn't think about the block scoping issue. What you have now works, unless you also want to move the comment to the end of the previous line, i.e.:
```ruby
      matched_regexp_group = nil  # @tag_regexp_list can be an empty list.
```

Reply:
Changed.